packages/react-native/ReactCommon/react/nativemodule/core/platform/ios/ReactCommon/RCTTurboModule.mm

```diff
@@ -400,12 +459,52 @@ id convertJSIValueToObjCObject(jsi::Runtime &runtime, const jsi::Value &value, s
       TurboModulePerfLogger::asyncMethodCallExecutionStart(moduleName, methodNameStr.c_str(), asyncCallCounter);
     }
 
-    @try {
-      [inv invokeWithTarget:strongModule];
-    } @catch (NSException *exception) {
-      throw convertNSExceptionToJSError(runtime, exception);
-    } @finally {
-      [retainedObjectsForInvocation removeAllObjects];
+    NSError *caughtException = nil;
+    BOOL shouldCatchException = isSync || reject;
+    try {
+      @try {
+        [inv invokeWithTarget:strongModule];
+       } @catch (NSException *exception) {
+         caughtException = maybeCatchException(
+                     shouldCatchException, @{
+                       @"name": exception.name,
+                       NSLocalizedDescriptionKey: exception.reason,
+                       @"stackSymbols": exception.callStackSymbols,
+                       @"stackReturnAddresses": exception.callStackReturnAddresses,
+                     }, nil);
+       } @catch (NSError *error) {
+         caughtException = maybeCatchException(shouldCatchException, nil, error);
+       } @catch (NSString *errorMessage) {
+         caughtException = maybeCatchException(shouldCatchException, @{
+                      NSLocalizedDescriptionKey: errorMessage,
+                     }, nil);
+       } @catch (id e) {
+         caughtException = maybeCatchException(shouldCatchException, @{
+                      NSLocalizedDescriptionKey: @"Unknown Objective-C Object thrown.",
+                     }, nil);
+       } @finally {
+         [retainedObjectsForInvocation removeAllObjects];
+       }
```

Comment:
Can you move this a c++ `finally`? In the case of a C++ exception we should also do this.

Reply:
Actually, I'm not even sure why we do this at all, this NSArray goes out of scope after this call anyway and should already be destroyed.

Reply:
I'm not sure either, but since it was there before I left it here.

I moved it after the try catch since cpp doesn't have finally.